### ISSUE TITLE
Compiling recursive functions fails in nopython mode

### ISSUE THREAD

**Issue by pitrou** [numba/numba#507]
```
>>> @jit(int_(int_), nopython=True)
... def fac(n):
...     if n == 1:
...         return 1
...     else:
...         return n * fac(n - 1)
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "numba\decorators.py", line 128, in wrapper
    disp.compile(sig)
  File "numba\dispatcher.py", line 108, in compile
    flags=flags, locals=locs)
  File "numba\compiler.py", line 94, in compile_extra
    return_type, flags, locals)
  File "numba\compiler.py", line 120, in compile_bytecode
    locals)
  File "numba\compiler.py", line 283, in type_inference_stage
    infer.propagate()
  File "numba\typeinfer.py", line 318, in propagate
    self.constrains.propagate(self.context, self.typevars)
  File "numba\typeinfer.py", line 116, in propagate
    constrain(context, typevars)
  File "numba\typeinfer.py", line 168, in __call__
    self.resolve(context, typevars, fnty)
  File "numba\typeinfer.py", line 177, in resolve
    sig = context.resolve_function_type(fnty, args, ())
  File "numba\typing\context.py", line 51, in resolve_function_type
    fnobj = func.overloaded.compile(tuple(args))
  File "numba\dispatcher.py", line 108, in compile
    flags=flags, locals=locs)
  File "numba\compiler.py", line 94, in compile_extra
    return_type, flags, locals)
  File "numba\compiler.py", line 120, in compile_bytecode
    locals)
  File "numba\compiler.py", line 283, in type_inference_stage
    infer.propagate()
  File "numba\typeinfer.py", line 318, in propagate
    self.constrains.propagate(self.context, self.typevars)
  File "numba\typeinfer.py", line 116, in propagate
    constrain(context, typevars)
  File "numba\typeinfer.py", line 168, in __call__
    self.resolve(context, typevars, fnty)
  File "numba\typeinfer.py", line 180, in resolve
    raise TypingError(msg, loc=self.loc)
numba.typeinfer.TypingError: Undeclared Dispatcher(CPUOverloaded(<function fac at 0x07548B30>))(int32,)
File "<stdin>", line 6
```

**Comment by jakevdp** [comment]
In #991 @seibert mentioned that recursion is not a high priority because it's only really used in trivial things like fibonacci numbers.

I have a use case that I think would justify work on this: tree-based exact nearest neighbors queries. The scikit-learn BallTree/KDTree is great, but one issue is that users can't easily define their own distance metrics for use in them, because this must be done at the Cython level. It would be nice to have a numba implementation of the tree code, because then a user could JIT-compile their custom metric and use it for fast neighbors queries. The most straightforward tree query code uses a recursive algorithm, and so it currently can't be run quickly with Numba. It is of course possible to replace the recursion with a manual call stack, but I don't really want to do that :smile:

I hope that gives some motivation to fix the recursion issue!

**Comment by seibert** [comment]
@jakevdp: Thanks for the example!  I agree this is a compelling reason to support recursion.  Is this direct recursion?

After the May release of Numba, we are going to spend the next release working on refactoring some internals of Numba that have gotten overly complicated with the feature-blitz of the last 5 releases.  I'm hoping during this work, we can try to address some of the reasons why supporting recursion is complicated in Numba, and make it easier to implement.

**Comment by jakevdp** [comment]
It is direct recursion; you can take a look at the code in this gist: https://gist.github.com/jakevdp/5216193

**Comment by sklam** [comment]
Basic direct recursion works on master right now.  Indirect recursion will work with https://github.com/numba/numba/pull/2130 merged.

**Comment by Juanlu001** [comment]
#2130 was merged, can this and #1845 be closed?

**Comment by seibert** [comment]
Good catch!  Thanks.
